File: Proyecto2_1096917/Proyecto2_1096917/BinaryTree.cpp
#include "pch.h"
#include "BinaryTree.h"
#include <iostream>

using namespace std;
template<typename T>
BinaryTree<T>::BinaryTree()
{
	root = NULL;
	numberOfNodes = 0;
}

template<typename T>
int BinaryTree<T>::weight() { return numberOfNodes; }

template<typename T>
void BinaryTree<T>::add(T element, int key) {
	if (this->root == NULL) {
		root = new TreeNode<T>(element, NULL, NULL, NULL, key);
		numberOfNodes++;
	}
	else {
		addElement(root, element, key);
	}
}

template<typename T>
void BinaryTree<T>::addElement(TreeNode<T> *root, T element, int key) {
	//si el elemento es menor que el valor de root lo agrega a la izq. recursivamente
	if (key < root->getKey()) {
		if (root->getLeft() == NULL) {
			root->setLeft(new TreeNode<T>(element, root, NULL, NULL, key));
			numberOfNodes++;
		}
		else {
			addElement(root->getLeft(), element, key);
		}
	}
	else if (key > root->getKey()) {// si es mayor lo agrega al subarbol derecho. 
		if (root->getRight() == NULL) {
			root->setRight(new TreeNode<T>(element, root, NULL, NULL, key));
			numberOfNodes++;
		}
		else {
			addElement(root->getRight(), element, key);
		}
	}
	else { //los elementos son iguales
		//no hace nada.
	}
}

template<typename T>
bool BinaryTree<T>::elementExists(TreeNode<T> *root, int elementKey) {
	if (this->root == NULL) {
		return false;
	}
	else if (elementKey == root->getKey()) {
		return true;
	}
	else
	{
		if (elementKey < root->getKey())
		{
			return elementExists(root->getLeft(), elementKey);
		}
		else
		{
			return elementExists(root->getRight(), elementKey);
		}
	}
}

template<typename T>
int BinaryTree<T>::numberOfChildren(TreeNode<T> *root) {
	int count = 0;
	if (root->getLeft() != NULL)
		count++;
	if (root->getRight() != NULL)
		count++;
	return count;
}

template<typename T>
void BinaryTree<T>::preOrder(TreeNode<T> * root) {
	if (root != NULL)
	{
		cout << root->getKey() << " ";
		preOrder(root->getLeft());
		preOrder(root->getRight());
	}
}

template<typename T>
void BinaryTree<T>::postOrder(TreeNode<T> *root) {
	if (root != NULL)
	{
		postOrder(root->getLeft());
		postOrder(root->getRight());
		cout << root->getElement() << " ";
	}
}

template<typename T>
void BinaryTree<T>::inOrder(TreeNode<T> *root) {
	if (root != NULL)
	{
		inOrder(root->getLeft());
		cout << root->getElement() << " ";
		inOrder(root->getRight());
	}
}

template<typename T>
T BinaryTree<T>::remove(TreeNode<T> *root, int elementKey) {
	if (this->root == NULL) {
		return NULL;
	}
	else if (elementKey == root->getKey()) {
		if (numberOfChildren(root) == 0) {
			int aux = root->getElement();
			if (root == this->root) {
				this->root = NULL;
			}
			else {
				if (elementKey == root->getParent()->getLeft()->getKey()) {
					root->getParent()->setLeft(NULL);
					root = NULL;
				}
				else {
					root->getParent()->setRight(NULL);
					root = NULL;
				}
			}
			numberOfNodes--;
			return aux;
		}
		else if (numberOfChildren(root) == 1) {
			int aux = root->getElement();

			if (root == this->root) {
				if (root->getLeft() != NULL) {
					this->root = root->getLeft();
				}
				else {
					this->root = root->getRight();
				}
			}
			else {
				if (root->getParent()->getLeft() != NULL) {
					if (root->getLeft() != NULL) {
						root->getParent()->setLeft(root->getLeft());
					}
					else {
						root->getParent()->setLeft(root->getRight());
					}
				}
				else {
					if (root->getLeft() != NULL) {
						root->getParent()->setRight(root->getLeft());
					}
					else {
						root->getParent()->setRight(root->getRight());
					}
				}
			}
			numberOfNodes--;
			return aux;
		}
		else {// el nodo que sustituirá será el más derecho del subárbol izquierdo.
			TreeNode *next = root->getLeft();
			int aux = root->getElement();
			if (next->getRight() != NULL) {
				while (next->getRight() != NULL)
				{
					next = next->getRight();
				}
				root->setElement(next->getElement());
				TreeNode *father = next->getParent();
				father->setRight(NULL);
			}
			else {
				root->setElement(next->getElement());
				root->setLeft(NULL);
			}
			numberOfNodes--;
			return aux;
		}
	}
	else {
		if (elementKey < root->getKey()) {
			return remove(root->getLeft(), elementKey);
		}
		else {
			return remove(root->getRight(), elementKey);
		}

	}
}

template<typename T>
int BinaryTree<T>::treeDepth(TreeNode<T> *root) {
	if (!root) {
		return 0;
	}
	else {
		int leftDepth = treeDepth(root->getLeft());
		int rightDepth = treeDepth(root->getRight());
		if (leftDepth > rightDepth) {
			return leftDepth + 1;
		}
		else {
			return rightDepth + 1;
		}
	}
}

template<typename T>
BinaryTree<T>::~BinaryTree()
{
}

File: ExamenFinalPA2018_1096917/DynamicStructures/EstrcuturasDinamicas/BinaryTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EstrcuturasDinamicas
{
    public class BinaryTree<T>
    {
        public TreeNode<T> root;
        private int numberOfTreeNodes = 0;
        Comparer<T> comp = Comparer<T>.Default;
        public MyLinkedList<TreeNode<T>> order = new MyLinkedList<TreeNode<T>>();
        public MyLinkedList<TreeNode<T>> repeated = new MyLinkedList<TreeNode<T>>();
        public MyLinkedList<int> repeatedKeys = new MyLinkedList<int>();
        Random generator = new Random();
        public delegate bool exists<T>(Node<TreeNode<T>> element1, Node<TreeNode<T>> element2);

        public BinaryTree()
        {
            root = null;
        }

        public int weight()
        {
            return numberOfTreeNodes;
        }


        public int treeDepth(TreeNode<T> root)
        {
            if (root != null)
            {
                return 0;
            }
            else
            {
                int leftDepth = treeDepth(root.getLeft());
                int rightDepth = treeDepth(root.getRight());
                return Math.Max(leftDepth, rightDepth);
            }
        }

        public void add(T element, int key)
        {
            if (this.root == null)
            {
                root = new TreeNode<T>(element, null, null, null, key);
                numberOfTreeNodes++;
            }
            else
            {
                addElement(root, element, key);
            }
        }

        private void addElement(TreeNode<T> root, T element, int key)
        {
            if (this.root == null)
            {
                this.root = new TreeNode<T>(element, null, null, null, key);
                numberOfTreeNodes++;
            }
            else
            {
                if (key < root.getKey())//x es menor que y
                {
                    if (root.getLeft() == null)
                    {
                        root.setLeft(new TreeNode<T>(element, root, null, null, key));
                        numberOfTreeNodes++;
                    }
                    else
                    {
                        addElement(root.getLeft(), element, key);
                    }
                }
                else if (key > root.getKey())//x es mayor que y 
                {
                    if (root.getRight() == null)
                    {
                        root.setRight(new TreeNode<T>(element, root, null, null, key));
                        numberOfTreeNodes++;
                    }
                    else
                    {
                        addElement(root.getRight(), element, key);
                    }
                }
                else//x es igual que y
                {
                    //si las llave actual es igual a la llave se le restará o se sumará un número aleatorio
                    //para luego agregarlo al arbol binario                    
                    repeatedKeys.addLast(key);                   
                    int num = generator.Next(2, 6);
                    key *= num;
                    add(element, key);
                }
            }
        }

        public MyLinkedList<int> repeatedID()
        {
            return repeatedKeys;
        }

        public MyLinkedList<TreeNode<T>> repeatedCustomers()
        {
            return repeated;
        }

        public bool elementExists(TreeNode<T> root, T element)
        {
            if (root == null)
            {
                return false;
            }
            else if (comp.Compare(element, root.getElement()) == 0)
            {
                return true;
            }
            else
            {
                if (comp.Compare(element, root.getElement()) < 0)
                {
                    return elementExists(root.getLeft(), element);
                }
                else
                {
                    return elementExists(root.getRight(), element);
                }
            }
        }

        public int numberOfChildren(TreeNode<T> root)
        {
            int count = 0;
            if (root.getLeft() != null)
                count++;
            if (root.getRight() != null)
                count++;
            return count;
        }

        public TreeNode<T> removeByKey(TreeNode<T> root, TreeNode<T> reference) {
            if (root == null)
            {
                return null;
            }
            else if (root.getKey() == reference.getKey())
            {
                if (numberOfChildren(root) == 0)
                {
                    TreeNode<T> aux = root;
                    if (root == this.root)
                    {
                        this.root = null;
                    }
                    else
                    {
                        if (reference.getKey() == root.getParent().getLeft().getKey())
                        {
                            root.getParent().setLeft(null);
                            root = null;
                        }
                        else
                        {
                            root.getParent().setRight(null);
                            root = null;
                        }
                    }
                    numberOfTreeNodes--;
                    return aux;
                }
                else if (numberOfChildren(root) == 1)
                {
                    TreeNode<T> aux = root;
                    if (root == this.root)
                    {
                        if (root.getLeft() != null)
                        {
                            this.root = root.getLeft();
                        }
                        else
                        {
                            this.root = root.getRight();
                        }
                    }
                    else
                    {
                        if (root.getParent().getLeft() != null)
                        {
                            if (root.getLeft() != null)
                            {
                                root.getParent().setLeft(root.getLeft());
                            }
                            else
                            {
                                root.getParent().setLeft(root.getRight());
                            }
                        }
                        else
                        {
                            if (root.getLeft() != null)
                            {
                                root.getParent().setRight(root.getLeft());
                            }
                            else
                            {
                                root.getParent().setRight(root.getRight());
                            }
                        }
                    }
                    numberOfTreeNodes--;
                    return aux;
                }
                else//El que sustituirá será el más derecho de los izquierdos
                {
                    TreeNode<T> next = root.getLeft();
                    TreeNode<T> aux = root;
                    if (next.getRight() != null)
                    {
                        while (next.getRight() != null)
                        {
                            next = next.getRight();
                        }
                        root.setElement(next.getElement());
                        TreeNode<T> father = next.getParent();
                        father.setRight(null);
                    }
                    else
                    {
                        root.setElement(next.getElement());
                        root.setLeft(null);
                    }

                    numberOfTreeNodes--;
                    return aux;
                }
            }
            else
            {
                if (reference.getKey() < root.getKey())
                {
                    return removeByKey(root.getLeft(), reference);
                }
                else
                {
                    return removeByKey(root.getRight(), reference);
                }
            }
        }

        public T remove(TreeNode<T> root, T element)
        {
            if (root == null)
            {
                return default(T);
            }
            else if (comp.Compare(element, root.getElement()) == 0)
            {
                if (numberOfChildren(root) == 0)
                {
                    T aux = root.getElement();
                    if (root == this.root)
                    {
                        this.root = null;
                    }
                    else
                    {
                        if (comp.Compare(element, root.getParent().getLeft().getElement()) == 0)
                        {
                            root.getParent().setLeft(null);
                            root = null;
                        }
                        else
                        {
                            root.getParent().setRight(null);
                            root = null;
                        }
                    }
                    numberOfTreeNodes--;
                    return aux;
                }
                else if (numberOfChildren(root) == 1)
                {
                    T aux = root.getElement();
                    if (root == this.root)
                    {
                        if (root.getLeft() != null)
                        {
                            this.root = root.getLeft();
                        }
                        else
                        {
                            this.root = root.getRight();
                        }
                    }
                    else
                    {
                        if (root.getParent().getLeft() != null)
                        {
                            if (root.getLeft() != null)
                            {
                                root.getParent().setLeft(root.getLeft());
                            }
                            else
                            {
                                root.getParent().setLeft(root.getRight());
                            }
                        }
                        else
                        {
                            if (root.getLeft() != null)
                            {
                                root.getParent().setRight(root.getLeft());
                            }
                            else
                            {
                                root.getParent().setRight(root.getRight());
                            }
                        }
                    }
                    numberOfTreeNodes--;
                    return aux;
                }
                else//El que sustituirá será el más derecho de los izquierdos
                {
                    TreeNode<T> next = root.getLeft();
                    T aux = root.getElement();
                    if (next.getRight() != null)
                    {
                        while (next.getRight() != null)
                        {
                            next = next.getRight();
                        }
                        root.setElement(next.getElement());
                        TreeNode<T> father = next.getParent();
                        father.setRight(null);
                    }
                    else
                    {
                        root.setElement(next.getElement());
                        root.setLeft(null);
                    }

                    numberOfTreeNodes--;
                    return aux;
                }
            }
            else
            {
                if (comp.Compare(element, root.getElement()) < 0)
                {
                    return remove(root.getLeft(), element);
                }
                else
                {
                    return remove(root.getRight(), element);
                }
            }
        }

        public void preOrder(TreeNode<T> root)
        {
            if (root != null)
            {
                Console.Write(root.getElement() + "->");
                preOrder(root.getLeft());
                preOrder(root.getRight());
            }
        }

        public void postOrder(TreeNode<T> root)
        {
            if (root != null)
            {
                postOrder(root.getLeft());
                postOrder(root.getRight());
                order.addLast(root);
            }
        }

        public void inOrder(TreeNode<T> root)
        {
            if (root != null)
            {
                inOrder(root.getLeft());
                order.addLast(root);
                inOrder(root.getRight());
            }
        }
    }
}
